### PR TITLE
feat(pageserver): use leases to temporarily block gc

### LINKED ISSUE #8063: Implement real lease logic: some map of LSNs to times in Timeline, plus delaying GC by the lease period on startup.
Part of #7497 

The pageserver will maintain a LSN to lease mapping and use it along with the other fields in `GCInfo` to decide which layers should be kept during GC.

### PR DESCRIPTION
**Pull request by yliang412**
Part of #7497, extracts from #7996, closes #8063.

## Problem

With the LSN lease API introduced in https://github.com/neondatabase/neon/issues/7808, we want to implement the real lease logic to prevent GC from proceeding pass some LSN when garbage collecting layers in a timeline.

To do so, we keeps an additional in-memory mapping of LSNs to leases in `GCInfo`. This mapping is updated when
- a request is made to grant a new lease (insertion of new leases/renewal of existing leases)
- `GCInfo` is refreshed during GC (removal of expired leases)

We use this mapping during GC similar to how we use `retain_lsns` for branches. The idea is that we will keep a layer if the layer start LSN less or equal to any of the leased LSN. This guarantees that we will keep all the layers needed to reconstruct all pages at all the leased LSNs with valid leases at a given time. 

### Future Task
- [ ] For the current lease implementation, we are actually keeping all the layers below the maximum LSN with valid leases (same problem for `retain_lsns`). Theoretically, for each LSN with valid lease, we only need to keep down to the most recent image layer at that lease/branch LSN. This can be a win for reducing the amount of data we retain.

## Summary of changes

- Maintain a in-memory mapping of LSN to lease expiration time in `GCInfo`.
- Requesting a lease for a LSN below `latest_gc_cutoff_lsn` will error out.
- Remove expired lease from the map when refreshing `GCInfo`.
- Block GC (similar to how we uses `retain_lsns`).
- A unit test test the `make_lsn_lease` API and how leases get used during GC.
- Delay GC by lease period at startup.
- Lease duration config in `TenantConf` (useful for running tests)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by skyzh — APPROVED** [review]

File: libs/pageserver_api/src/models.rs [hunk 1]
```diff
@@ -177,6 +177,28 @@ serde_with::serde_conv!(
     |value: String| -> Result<_, humantime::TimestampError> { humantime::parse_rfc3339(&value) }
 );
 
+impl LsnLease {
+    /// The default length for an explicit LSN lease request (10 minutes).
+    #[cfg(not(feature = "testing"))]
```
Comment (yliang412) [inline libs/pageserver_api/src/models.rs:182]: I feel we want to shorten the lease interval for testing. Not sure what are the best practice for using feature flags.

File: pageserver/src/tenant/tasks.rs [hunk 2]
```diff
@@ -362,6 +364,11 @@ async fn gc_loop(tenant: Arc<Tenant>, cancel: CancellationToken) {
 
             if first {
                 first = false;
+
+                if delay_by_default_lease_length(&cancel).await.is_err() {
```
Comment (yliang412) [inline pageserver/src/tenant/tasks.rs:368]: delay is made cancellable.

File: libs/pageserver_api/src/models.rs (hunk 1 above)
Comment (skyzh) [reply libs/pageserver_api/src/models.rs:182]: This looks fine. I think this is also possible with const fns if you feel that's a better way to do that?

Comment (yliang412) [reply libs/pageserver_api/src/models.rs:182]: I'm actually trying to put these inside `TenantConf` based on Conrad's suggestion so we can have a runtime config? Does that sounds good?

File: pageserver/src/tenant/timeline.rs [hunk 3]
```diff
@@ -2117,6 +2117,24 @@ const REPARTITION_FREQ_IN_CHECKPOINT_DISTANCE: u64 = 10;
...
+    }
+
+    // TODO(yuchen): remove unused flag after implementing https://github.com/neondatabase/neon/issues/8072
+    #[allow(unused)]
+    pub(crate) fn get_lsn_lease_length_for_ts(&self) -> Duration {
+        let tenant_conf = self.tenant_conf.load();
+        tenant_conf
+            .tenant_conf
+            .lsn_lease_length
+            .unwrap_or(self.conf.default_tenant_conf.lsn_lease_length)
```
Comment (skyzh) [inline pageserver/src/tenant/timeline.rs:2135]: should be lsn_lease_length_for_ts instead?

**Review by skyzh — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs (hunk 3 above)
Comment (yliang412) [reply pageserver/src/tenant/timeline.rs]: addressed

**Pull request merged**